### ISSUE TITLE
OpenNDS randomly crashed

### ISSUE THREAD

**Issue by minhng99** [openNDS/openNDS#165]
Hi, I've been stuck on openNDS v7.0.1 with the `iptables_fw_destroy` patch for a while because something on the newer version just keeps crashing, now I've finally had some time to investigate it.

Attached is the syslog, filtered to only display opennds-related logs, opennds is running with debuglevel 3 and is on the latest OpenWRT 19.07

I've found that whenever this happen then it will just crash opennds
```

Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: preauthenticated: host [(null)] url [/blank.html]
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: url is [ /blank.html ], checking for [ opennds_preauth ]
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: url is [ /blank.html ], checking for [ opennds_deny ]
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: preauthenticated: Requested Host is [ (null) ]
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: preauthenticated: Requested url is [ /blank.html ]
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: preauthenticated: FAS is enabled
Thu May 20 08:10:47 2021 daemon.debug opennds[31255]: url is [ /blank.html ], checking for [ opennds_auth ]

```
[log_opennds.txt](https://github.com/openNDS/openNDS/files/6514404/log_opennds.txt)

**Comment by bluewavenet** [comment]
@minhng99 
Sorry I'm not following what you are saying.. I probably need another coffee :-)

> openNDS v7.0.1 with the iptables_fw_destroy patch

What is that?

> something on the newer version just keeps crashing

The current release version is 9.0.0, do you mean 9.0.0, or 8.x.x?

**Comment by minhng99** [comment]
Oh I mean I'm currently running v7.0.1 with the cherry-pick for the fix that I've referenced in https://github.com/openNDS/openNDS/issues/96 (sorry I couldn't find the specific commit)

the log above is from v9.0.0, I also have some weird crashes back when I try to use version 8.x but I don't have time to debug it so I think this crash has something to do with whatever is implemented after the v7.0.1 which I'm currently running just fine

**Comment by bluewavenet** [comment]
@minhng99 
Because host is null, it crashes. But host should not be able to be null......( there should be a trap in the code rather than crashing...)
Question is how is it getting redirected to here with no host....?
OH! you have fasfqdn set to `wificp.testcase.vn` and this resolves to null (unless you have added it to hosts or dnsmasq)

**Comment by minhng99** [comment]
> @minhng99
> Because host is null, it crashes. But host should not be able to be null......( there should be a trap in the code rather than crashing...)
> Question is how is it getting redirected to here with no host....?
> OH! you have fasfqdn set to `wificp.testcase.vn` and this resolves to null (unless you have added it to hosts or dnsmasq)

Yes, we have a local DNS which can resolves the `wificp.testcase.vn` to a proper IP

I'm using opennds to authenticate wifi guests by a weird method :P 

I have a FAS server that displays an HTML page with a button that will POST data to the AP's webserver cgi-bin script, that script will do some check then it will call `ndsctl auth` the client MAC in that POST data

Also, I have a script hooked in `dnsmasq` to auth and allow client to access the internet immediately if they have already authenticated but roams to another AP

**Comment by bluewavenet** [comment]
@minhng99 
I'm trying to replicate the problem....
It is something to do with the way you are calling the auth, but it should be ok...
I guess it is an edge case bug ;-)
I'll get that cup of coffee and look closer!

**Comment by bluewavenet** [comment]
@minhng99 

> I have a FAS server that displays an HTML page with a button that will POST data to the AP's webserver cgi-bin script, that script will do some check then it will call ndsctl auth the client MAC in that POST data

I like it!

**Comment by minhng99** [comment]
> I guess it is an edge case bug ;-)

Yeah... I don't know... but everything seems to work fine on v7.0.1, I don't know if its the same bug but I also getting weird crashes on v8.x so if this issue is related then I think the issue started somewhere around that time.

I'm only tried v9.0 a few days ago because the random `gatewaymac` being pushed to my FAS server and I thought maybe it has been fixed in v9.0, I still don't know why it happened and still having that issue, a lot of AP suddenly submitting random MAC in 1 particular day and it's still go up, this issue never has happened before

**Comment by bluewavenet** [comment]
@minhng99 

> Thu May 20 08:10:05 2021 daemon.debug opennds[28997]: preauthenticated: Requested Host is [ (null) ]
> Thu May 20 08:10:05 2021 daemon.debug opennds[28997]: preauthenticated: Requested url is [ /blank.html ]

Where is /blank.html? Somewhere on your FAS?
It looks like openNDS is being requested for http://(something)/blank.html  but MHD is returning (null), which it should never do.
So it looks like a bug in MHD, followed by openNDS not checking the sanity of MHD's response....
It is simple to fix the crash, but what are you doing to hit the (probable) bug in MHD?

**Comment by minhng99** [comment]
I don't use the MHD port of the opennds at all, all it does is to do a CPD pop up client to the FAS server then I authenticate it via the cgi-bin script method, I think this I what the client has requested but I'm not sure why.

**Comment by bluewavenet** [comment]
@minhng99 
The main question:
 * Where is blank.html?

**_A separate issue_** (depending on the details of what you are actually doing):

> I have a FAS server that displays an HTML page with a button that will POST data to the AP's webserver cgi-bin script, that script will do some check then it will call ndsctl auth the client MAC in that POST data

You should generate rhid from hid and the preshared key then authenticate using: 
`ndsctl auth <rhid>`

Otherwise anyone could [potentially] bypass your client validation by POST of their mac to the cgi-bin script

**Comment by minhng99** [comment]
> @minhng99
> The main question:
> 
> * Where is blank.html?
 Nowhere, my entire system from the AP to the FAS server don't have any reference to any file named like that, I was surprised when I come across this in the debug log so I've done a grep search but nothing, not even in the opennds source code.

> Otherwise anyone could [potentially] bypass your client validation by POST of their mac to the cgi-bin script

Yeah I know, it's a low security thing anyway, if someone really wanted to bypass an advertising page then they could just POST their own MAC to my cgi-bin script which is not even encrypted, but I don't think most of my guests know how to do that :)

**Comment by bluewavenet** [comment]
@minhng99 
I have merged a fix for the crash into master.
#166  Fix - edge case where MHD returns (null) as host value
This does not fix the root cause - It seems to be a an artefact of the way you are handling authentication, but I do not know what.

With this fix, openNDS will generate an error 511 instead of crashing.

If you could test this, I would much appreciate it.

**Comment by minhng99** [comment]
Okay... I've run the latest version for 3 days and there seems to be no crash yet, I'm still getting `client access: GET /blank.html` in the log with other garbage paths that I don't think is existent on the MHD server, I think these things are just what the client device requested automatically...

But I've found that opennds has opened a lot of thread on the port 2050, is this supposed to happen? there are ~500 of these...
![image](https://user-images.githubusercontent.com/14113155/119284901-00732200-bc6b-11eb-8f4d-b3d9f99ef72b.png)
![image](https://user-images.githubusercontent.com/14113155/119284974-30bac080-bc6b-11eb-8da5-859396a2680d.png)

ndsctl is locked up and is no longer usable.

I've got another syslog for you here:

[opennds-new.txt](https://github.com/openNDS/openNDS/files/6529325/opennds-new.txt)

**Comment by bluewavenet** [comment]
@minhng99 
Excellent information thanks. It will take a while to digest it all ;-)

Some things that immediately come to light:
 * There are many POST requests eg:
     * `POST /dc/fcompb.pgs`
     * `POST /qq_product_operations/nettest/index2.html`
     * `POST /notification_center/1bd9140a-71e7-4ea7-a3b9-d4f7ebe2eafc/1621602777152`
     * `POST /post/Query_Update`
     *  etc. These can only happen with a direct request to MHD ie not as a result of the port 80 redirect.
 *  There seem to be many clients in the client list that are not connected eg:
     * `Failed to get client connections for [08:78:08:45:93:14] - giving up`
     * `Failed to get client connections for [72:68:8e:de:91:ea] - giving up`
     * etc. This normally only happens when clients move out of range but have not been purged yet by preauthidletimout. There seems to be rather a lot of them for a test system......
 * All this happens over and over 
 * Eventually openNDS seems to run out of file descriptors ie:
     * `popen(): [No file descriptors available] Giving up..`
 * At this point ndsctl will go into a wait loop waiting for an unlock

This shows there may be a bug in openNDS where file descriptors are not being freed - I will look into this.

**However this is not the root cause of your problem,** that must be _somewhere in your FAS/cgi method_ as indicated by the POST errors for example.

If you could review your FAs/cgi code with this in mind, it would be great and I will look at the file descriptors thing.

**Comment by minhng99** [comment]
> These can only happen with a direct request to MHD ie not as a result of the port 80 redirect.
I don't know why these requests are being made, I don't have such URL on any of my system, the URL looks like some kind of captive portal detection on the client device, none of the devices should've directly interacted with the MHD port for authentication.

> There seems to be rather a lot of them for a test system
I have this AP testing running at my lab but having the same SSID as the production ones currently running in the building so maybe some devices have automatically connected to it.

> If you could review your FAs/cgi code with this in mind, it would be great and I will look at the file descriptors thing.
No idea, the cgi code just doing some statistic logging using CURL to an HTTP endpoint, write the client MAC address into my radius server then calling `ndsctl auth [MAC]` and nothing else, I don't know what could've caused this.

**Comment by bluewavenet** [comment]
@minhng99 

> the URL looks like some kind of captive portal detection on the client device

I'll look into this closer but it looks like it might be crypto currency/blockchain mining attempts combined with the use of SniffAir and NetTest hacker tools......

**Comment by bluewavenet** [comment]
6:52 am for me - I need a coffee :-D

**Comment by bluewavenet** [comment]
@minhng99 

> the cgi code just doing ....... then calling ndsctl auth [MAC]

Probably not related to the issue here, but instead of calling ndsctl directly, it would be better to use the functionality of the do_ndsctl() function in the library script /usr/lib/opennds/libopennds.sh
(A useful enhancement for the next release would be to make this an available call to libopennds... )

**Comment by minhng99** [comment]
> @minhng99
> 
> > the cgi code just doing ....... then calling ndsctl auth [MAC]
> 
> Probably not related to the issue here, but instead of calling ndsctl directly, it would be better to use the functionality of the do_ndsctl() function in the library script /usr/lib/opennds/libopennds.sh
> (A useful enhancement for the next release would be to make this an available call to libopennds... )


Hmm... apparently that function is nowhere to be found in the v7.0.1 which I'm currently using in production, for some reason only that v7.0.1 is the most stable version there is for me...

**Comment by bluewavenet** [comment]
@minhng99 

> that function is nowhere to be found in the v7.0.1

No not in v7.0.1 - it's in libopennds.sh in v9.0.x so you should have it on your test system.
You would have to copy the code and use it in your cgi script where you want to call ndsctl.

**Comment by bluewavenet** [comment]
@minhng99 
Did you identify where the hacker tool probing was coming from?

**Comment by minhng99** [comment]
No idea, I don't think someone is trying to hack it because it's a guest WiFi system and it's not even psk'ed

**Comment by bluewavenet** [comment]
@minhng99  
`/dc/fcompb.pgs` is used by BscScan blockchain scanner, so someone is Crypto mining.

**Comment by minhng99** [comment]
Interesting... probably someone in the office is using their company PC for mining coins

**Comment by bluewavenet** [comment]
@minhng99 
`POST /x` is an initial php vulnerability test often used by hacker bots as a staring point. MHD ignores it so the bot goes away.

**Comment by bluewavenet** [comment]
`POST /parse/classes/_Installation` is often seen when bots probe for node.js vulnerabilities

**Comment by minhng99** [comment]
maybe it's a sketchy random Chinese android phone being used as a botnet?

**Comment by bluewavenet** [comment]
`POST /mmtls/` is a probe for tls mutual authentication exchange vulnerability.

> maybe it's a sketchy random Chinese android phone being used as a botnet?

Whatever it is / they are, it may well be the root cause of your problems .....
It is not helpful for you at least directly, _but it is great for testing openNDS_ :-D
I might have found where it is not freeing file descriptors.... just going to test.

**Comment by bluewavenet** [comment]
@minhng99 

> I might have found where it is not freeing file descriptors.... just going to test.

When I looked closer I could not see anything wrong.

I am not able to replicate the problem after lots of testing....
In your case, openNDS stopped responding because of:
`popen(): [No file descriptors available] Giving up..`

Try installing the `lsof` package and see if you can find out what is happening.

**Comment by minhng99** [comment]
Yeah... unfortunately I've installed a different firmware on my test AP, going to install the 9.0.1b to see if the same issue happens again, this time I'll get the lsof output for you.

**Comment by minhng99** [comment]
And... it's happened again after just 2 hours of uptime:
![image](https://user-images.githubusercontent.com/14113155/119802316-a8594b80-bf08-11eb-9f76-a8a1c0f6f228.png)

I've attached the syslog and lsof below.
Thanks!
[opennds-may27.txt](https://github.com/openNDS/openNDS/files/6552761/opennds-may27.txt)

[lsof-may27.txt](https://github.com/openNDS/openNDS/files/6552760/lsof-may27.txt)

**Comment by bluewavenet** [comment]
@minhng99 
It seems from your logs that MHD is in a constant failing state. I am still unable to replicate this.
I have however added an improved MHD watchdog and merged into master.
If you could try this, it would be much appreciated. There is some extra debug logging that might help....

**Comment by minhng99** [comment]
Got it, will build the master branch now

**Comment by minhng99** [comment]
It doesn't seem to happen anymore after 3 hours of uptime... but I can see MHD is constantly crashing in the log.
[lsof-may28.txt](https://github.com/openNDS/openNDS/files/6558627/lsof-may28.txt)
[opennds-may28.txt](https://github.com/openNDS/openNDS/files/6558628/opennds-may28.txt)

**Comment by bluewavenet** [comment]
@minhng99 

> It doesn't seem to happen anymore after 3 hours of uptime.

Yes, it looks like the improved watchdog frees up old file descriptors used by dead MHD instances. 

> but I can see MHD is constantly crashing in the log

When it is in this state:
 * can new clients log in?
 * on a client browser, what does `http://status.client `display?

Have you tried a test setup NOT using your FAS/cgi, ie using one of the the default modes? I have tested with default modes, supplied FAS examples and have 9.0.0 in production with custom FAS with no sign of this problem.
I am sure that when we find it this will be a simple issue ;-)

> If you want I could give you SSH access to that device

Might speed up the debugging...

**Comment by bluewavenet** [comment]
>         Have you tried a test setup NOT using your FAS/cgi
> 
> Well... MHD seems to be crashing even when there's no one logging into it
> 

Yes but have you tried a test setup WITHOUT your FAS/cgi?

>         on a client browser, what does http://status.client display?
> 
> nothing, it doesn't resolve this domain

Do you have it disabled ie, in the config:
`option gatewayfqdn 'disable'`

If not then this might be significant.

What port is your cgi server using?

**Comment by minhng99** [comment]
> Yes but have you tried a test setup WITHOUT your FAS/cgi?

Uh... no... I'll try this next week

> Do you have it disabled ie, in the config:

No, this is what I have in my configs

```
config opennds 'opennds1'
        list authenticated_users 'allow all'
        option enabled '1'
        option fwhook_enabled '1'
        option maxclients '250'
        option authidletimeout '60'
        option fas_secure_enabled '2'
        option faskey '0a28a5d478967f0aecb9b31466c0a067'
        option faspath '/net.wifi-client-script/v2/index.html'
        option fasport '9988'
        option fasremotefqdn 'wificp.fpt.vn'
        option sessiontimeout '60'
        option debuglevel '3'
        list users_to_router 'allow tcp port 53'
        list users_to_router 'allow tcp port 22'
        list users_to_router 'allow tcp port 23'
        list users_to_router 'allow tcp port 80'
        list users_to_router 'allow tcp port 443'
        list users_to_router 'allow tcp port 1999'
        list users_to_router 'allow udp port 53'
        list users_to_router 'allow udp port 67'
        list walledgarden_fqdn_list 'fptplay.tv'
        list walledgarden_fqdn_list 'st88.adsplay.net'
        list walledgarden_fqdn_list 'cdn.adsplay.xyz'
        list walledgarden_fqdn_list 'cdn.adsplay.net'
        list walledgarden_fqdn_list 'd.adsplay.net'
        list walledgarden_fqdn_list 'd8.adsplay.net'
        list walledgarden_fqdn_list 'fpt.vn'
        list walledgarden_port_list '80'
        list walledgarden_port_list '443'
        list walledgarden_port_list '9988'

```

> What port is your cgi server using?

`:1999`, I just create another port using the pre-installed uhttpd for the Luci panel and put my cgi-bin script in `/www/cgi-bin/`

**Comment by minhng99** [comment]
Is there any how that external changes that made to the `walledgarden` ipset have anything to do with this?

I have an external script which do dns resolve a list of domain then add the IP to the walledgarden ipset, I do this because there seems to be some issue with the dnsmasq ipset which always duplicate its list whenever I rebooted.

So the walledgarden list I have in opennds config contains very little but I do have the need to add a lot of domain which bugged something.

**Comment by bluewavenet** [comment]
> Is there any how that external changes that made to the walledgarden ipset have anything to do with this?

Depends on what those changes are, but I don't think so.

> there seems to be some issue with the dnsmasq ipset which always duplicate its list whenever I rebooted.

I don't see that problem. On startup, openNDS will delete the ipset "walledgarden" if it exists then re-create it. It will add the ipset to dnsmasq config if it is not already there.
Now, when clients do DNS queries for FQDNs in "walledgarden", dnsmasq will add the ipaddresses it resolves to the ipset.

From your logs:
```
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Executing command: ipset -v
Thu May 27 07:10:46 2021 daemon.notice opennds[5248]: ipset support is available
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Executing command: dnsmasq --version | grep ' ipset '
Thu May 27 07:10:46 2021 daemon.notice opennds[5248]: dnsmasq ipset support is available
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Executing command: ipset destroy walledgarden
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Executing command: ipset create walledgarden hash:ip
Thu May 27 07:10:46 2021 daemon.info opennds[5248]: Walled Garden ipset created
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Dnsmasq Walled Garden config [/fpt.vn/d8.adsplay.net/d.adsplay.net/cdn.adsplay.net/cdn.adsplay.xyz/st88.adsplay.net/fptplay.tv/walledgarden]
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: Executing command: /usr/lib/opennds/dnsconfig.sh "ipsetconf" "/fpt.vn/d8.adsplay.net/d.adsplay.net/cdn.adsplay.net/cdn.adsplay.xyz/st88.adsplay.net/fptplay.tv/walledgarden"
Thu May 27 07:10:46 2021 daemon.info opennds[5248]: Dnsmasq configured for Walled Garden
Thu May 27 07:10:46 2021 daemon.debug opennds[5248]: restart command [ /usr/lib/opennds/dnsconfig.sh "restart_only" & ]
Thu May 27 07:10:46 2021 daemon.info opennds[5248]: Dnsmasq restarted

```

**Comment by minhng99** [comment]
Yeah, I know that, but it's seems to be bugged out when I have ~60 domain in opennds's walledgarden list, the thing split the ipset entry in the dnsmasq into two list then it bugged out and keep adding more and more each reboot, making dnsmasq failed to start.

**Comment by bluewavenet** [comment]
> ~60 domain in opennds's walledgarden list

Wow!! Why? (why not I suppose)
The walled garden is intended purely to allow access to things required during the login process, so 60 FQDNs seems a bit excessive given that most client CPD processes restrict what can be displayed.
It would be sensible, if you really need this number, to put them in their own ipset, or is that what you ended up doing?

The openNDS dnsconfig.sh library may well be failing with such a long list, I will have to take a look at it. It has been tested with 6 if I remember correctly. Usual configs have 2 or 3.

**Comment by minhng99** [comment]
Yeah.... I'm operating a pretty large amount of public Internet AP (~5000 AP) and our advertisement partner wants the captive portal to redirects the user to their  website *after* the user have clicked the "Authenticate" button

This is pretty tricky to do because they don't allow us to modify their site... but then most of CPD pop-up will close immediately after it's realised that it's having Internet access, what we've done is to delay the `ndsctl auth` calling in the cgi-bin for a few seconds to please our advertising partner.

And we have to add their site and every resource domain that their site needs to display properly, this leads to a pretty big walledgarden list

Guess I'm really an edge case user, lol

**Comment by bluewavenet** [comment]
Probably a very good idea to have a separate ipset for the advertising partner. 
I'm not sure dnsmasq is designed to cope with that many entries either.
You need an inverse adblock!

> what we've done is to delay the ndsctl auth

The problem with that is if you delay too long the client device CPD will timeout and jump to any other ssid it knows about.

**Comment by minhng99** [comment]
> You need an inverse adblock!

At first we thought we would need to block every common CPD domain while still let client having Internet access, most people will think they needs to go through that before they can access the Internet so it would buy us some time but then while trying to reverse-engineering opennds without reading its code, we realise that all it does is to add the domain into the ipset so our current workaround is working just fine, I'm currently having ~100 IP in the ipset and I'm surprised that it's still work, heh.

**Comment by bluewavenet** [comment]
Hundreds of ips in an ipset is not a problem. Hundreds of fqdns in dnsmasq is a different issue!

**Comment by minhng99** [comment]
Well.. we did thought of the case if some domain have multiple IP in its A record then there's a chance that someone will resolve that domain into an IP that we don't have in the ipset thus make it inaccessible... but no one has complained about this so it's probably a small percentage, lol

We updates the domain list once in a few days but there's a crontab that does refresh the ipset once every 15 minutes to make sure the ipset contains all the latest IP record of those multi-IP domains.

**Comment by bluewavenet** [comment]
It works - that is what matters!

We just need to get to the bottom of the MHD crashing.
First I would suggest a quick test with:
`option gatewayfqdn 'disable'`
`option fasremoteip '118.68.13.46'`

**Comment by bluewavenet** [comment]
@minhng99 
I have it! I think..
Tell me, you have the full version of wget installed don't you?
I replicated the MHD issue by installing wget full version. 

I have added my fix and will push to master......

**Comment by minhng99** [comment]
Oh yes, I do have wget-full installed, what does wget have anything to do with....

Is it because your MHD check script is expecting something? Damn

**Comment by bluewavenet** [comment]
@minhng99 
_I did say it would most likely be something simple ;-)_

Yes, the MHD watchdog uses wget to test and expects a specific answer.

In systems with limited free RAM, under high load, MHD starts to return errors when it cannot allocate memory. To begin with it recovers nicely, but as the load increases even more it eventually becomes unresponsive. The watchdog catches this and restarts it. This was intermittently seen in production on routers with 64MB RAM with >100 clients and was pretty hard to find!

On your setup, the watchdog was interpreting its check as MHD failed so was continually restarting it.

So now we have a much improved watchdog catering for many edge cases :-D

**Thankyou for your efforts.**

I will merge into master shortly and would appreciate it if you could try this one out!

With regards to `http://status.client` this needs further investigation at your end as it indicates some other problem.
**This is concerning as the MHD watchdog uses _part_ of this functionality.**
We should start with adding:
`option gatewayfqdn 'disable'`
`option fasremoteip '118.68.13.46'`
and verifying everything is still working with your FAS/cgi.

**Comment by minhng99** [comment]
Yeah I'll try it on Monday because I left the AP in my office, many thanks!

And... I'm still having issue with the random MAC issue, still have no idea where it comes from but I've changed the `gatewaymac` in the code so it just copy from the `gatewayname` which I've inserted using the init.d config generator.

**Comment by bluewavenet** [comment]
@minhng99 
Excellent.
Your roaming is going to be problematic as mac randomisation looks like it is the default for the latest versions of iOS and Android. Not much we can do about it....

**Comment by minhng99** [comment]
Yeah... not much I could do about it either... but looks like it's only going to change MAC address if you "forget" an SSID, other than that it doesn't change on my phone

**Comment by bluewavenet** [comment]
Mmm, on the same AP yes, but the same SSID on a different AP, no. It seems to look at the BSS (ie mac of the wireless), at least that is what seems to happen with my phone..... not sure.
